### ISSUE TITLE
[Serializer][ORM] enable_lazy_ghost_objects with Serializer doesn't work

### ISSUE THREAD

**Issue by Hanmac** [symfony/symfony#53972]
### Symfony version(s) affected

5.4.34

### Description

enable_lazy_ghost_objects is set to true

have a Class with Complex Structure and Collections

```php

use Doctrine\Common\Collections\Collection;
interface PageInterface
{
    /**
     * @return Collection<array-key, PageInterface>
     */
    public function getChildren(): Collection;
    /**
     * @return Collection<array-key, BlockInterface>
     */
    public function getBlocks(): Collection;
}

```

```php

use Doctrine\Common\Collections\Collection;
interface BlockInterface
{
    /**
     * @return Collection<int, BlockInterface>
     */
    public function getChildren(): Collection;
}
```

### How to reproduce

```yaml
doctrine:
    dbal:
        url: '%env(resolve:DATABASE_URL)%'

        # IMPORTANT: You MUST configure your server version,
        # either here or in the DATABASE_URL env var (see .env file)
        #server_version: '13'
    orm:
        auto_generate_proxy_classes: true
        enable_lazy_ghost_objects: true
```

have entity classes that implement this interfaces:

```php

use Doctrine\Common\Collections\Collection;
interface PageInterface
{
    /**
     * @return Collection<array-key, PageInterface>
     */
    public function getChildren(): Collection;
    /**
     * @return Collection<array-key, BlockInterface>
     */
    public function getBlocks(): Collection;
}

```

```php

use Doctrine\Common\Collections\Collection;
interface BlockInterface
{
    /**
     * @return Collection<int, BlockInterface>
     */
    public function getChildren(): Collection;
}
```

now use Symfony Normalizer to try to normalize the Page object when the Collections aren't fully loaded yet

This ErrorException happened:
```
ErrorException:
User Notice: Undefined property: Proxies\__CG__\App\Entity\SonataPageBlock::$lazyObjectState in D:\*****\******\******\sonata\private\vendor\symfony\property-access\PropertyAccessor.php on line 488

  at D:\*****\******\******\sonata\private\vendor\symfony\var-exporter\LazyGhostTrait.php:193
  at Proxies\__CG__\App\Entity\SonataPageBlock->__get('lazyObjectState')
     (D:\*****\******\******\sonata\private\vendor\symfony\property-access\PropertyAccessor.php:488)
  at Symfony\Component\PropertyAccess\PropertyAccessor->readProperty(array(object(SonataPageBlock)), 'lazyObjectState', false)
     (D:\*****\******\******\sonata\private\vendor\symfony\property-access\PropertyAccessor.php:154)
  at Symfony\Component\PropertyAccess\PropertyAccessor->getValue(object(SonataPageBlock), 'lazyObjectState')
     (D:\*****\******\******\sonata\private\vendor\symfony\serializer\Normalizer\ObjectNormalizer.php:136)
  at Symfony\Component\Serializer\Normalizer\ObjectNormalizer->getAttributeValue(object(SonataPageBlock), 'lazyObjectState', null, array('datetime_format' => 'U', 'skip_null_values' => true, 'callbacks' => array('blocks' => object(Closure), 'parent' => object(Closure)), 'cache_key' => false, 'circular_reference_limit_counters' => array('00000000000005400000000000000000' => 1, '00000000000008a70000000000000000' => 1)))
     (D:\*****\******\******\sonata\private\vendor\symfony\serializer\Normalizer\AbstractObjectNormalizer.php:190)
  at Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer->normalize(object(SonataPageBlock), null, array('datetime_format' => 'U', 'skip_null_values' => true, 'callbacks' => array('blocks' => object(Closure), 'parent' => object(Closure)), 'cache_key' => false, 'circular_reference_limit_counters' => array('00000000000005400000000000000000' => 1, '00000000000008a70000000000000000' => 1)))
     (D:\*****\******\******\sonata\private\vendor\symfony\serializer\Serializer.php:161)
  at Symfony\Component\Serializer\Serializer->normalize(object(SonataPageBlock), null, array('datetime_format' => 'U', 'skip_null_values' => true, 'callbacks' => array('blocks' => object(Closure), 'parent' => object(Closure)), 'cache_key' => false, 'circular_reference_limit_counters' => array('00000000000005400000000000000000' => 1)))
     (D:\*****\******\******\sonata\private\vendor\symfony\serializer\Serializer.php:179)
  at Symfony\Component\Serializer\Serializer->normalize(array(object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock), object(SonataPageBlock)), null, array('datetime_format' => 'U', 'skip_null_values' => true, 'callbacks' => array('blocks' => object(Closure), 'parent' => object(Closure)), 'cache_key' => false, 'circular_reference_limit_counters' => array('00000000000005400000000000000000' => 1)))
     (D:\*****\******\******\sonata\private\vendor\symfony\serializer\Normalizer\AbstractObjectNormalizer.php:224)
  at Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer->normalize(object(SonataPagePage), null, array('datetime_format' => 'U', 'skip_null_values' => true, 'callbacks' => array('blocks' => object(Closure), 'parent' => object(Closure)), 'cache_key' => false, 'circular_reference_limit_counters' => array('00000000000005400000000000000000' => 1)))
     (D:\*****\******\******\sonata\private\vendor\symfony\serializer\Serializer.php:161)
  at Symfony\Component\Serializer\Serializer->normalize(object(SonataPagePage), null, array('datetime_format' => 'U', 'skip_null_values' => true, 'callbacks' => array('blocks' => object(Closure), 'parent' => object(Closure))))

```

### Possible Solution

calling `Collection->getValues()` might not be enough, the 'blocks' Closure already does it

```php
AbstractNormalizer::CALLBACKS => [
    'blocks' => static fn (Collection $collection, PageInterface $object, string $attribute, ?string $format = null, array $context = []) => $collection->filter(static fn (BlockInterface $block) => !$block->hasParent())->getValues(),
]
```           

### Additional Context

Related Sonata Issue:
https://github.com/sonata-project/SonataPageBundle/issues/1747

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Comment by Hanmac** [comment]
@haivala can you do a example application?

i don't know exactly what causes this problem

**Comment by craigh** [comment]
refs https://github.com/doctrine/orm/issues/11203

**Comment by haivala** [comment]
> @haivala can you do a example application?
> 
> i don't know exactly what causes this problem

You install SonataPageBundle, create site and page for it. Then you add block (for example TextBlock from SonataBlockBunle) to that page and try to create Snapshot of that page. That is when the error appears.

**Comment by xabbuh** [comment]
Can you create a small example application with the needed changes that allow someone to reproduce the issue?

**Comment by haivala** [comment]
> Can you create a small example application with the needed changes that allow someone to reproduce the issue?

I wish we had Sonata demo!

I'm unable to do example application as I'm sick right now. That said:
Just installing SonataPageBundle with the documentation and depending if `enable_lazy_ghost_objects` is set to true you get this error on Symfony 6.4 and 5.4.

**Comment by Hanmac** [comment]
> Can you create a small example application with the needed changes that allow someone to reproduce the issue?

The problem with the example application is, that it might depend on the database settings

for example i couldn't reproduce the error with unit-tests that uses sqlite (or maybe i used the wrong settings)

so an example application would probably need to use mysql/mariadb database

**Comment by xabbuh** [comment]
Without being able to reproduce the issue I am afraid that nobody will come up with a fix. So I am closing here for now, but we can consider to reopen when we are able to reproduce.

**Comment by Hanmac** [comment]
@xabbuh i can TRY to make a example application, but how are you going to test it anyway?
it would probably be need a mariadb for the doctrine part, and i don't know how to make this reproduceable

**Comment by xabbuh** [comment]
A Docker setup would probably be needed then.

**Comment by haivala** [comment]
I'll try to do it now

**Comment by haivala** [comment]
https://github.com/haivala/reproducer
Got it with SonataPageBundle `multisite: host_with_path_by_locale` and its runtime not the `multisite: host` without runtime

**Comment by nicolas-grekas** [comment]
Possibly fixed by #54194, please confirm if you can.
